src/server/serverGrouping.ts

```diff
@@ -0,0 +1,467 @@
+			new HttpServer(_extensionUri, _reporter, _endpointManager, _connection)
+		);
+
+		this._watcher = vscode.workspace.createFileSystemWatcher('**');
+
+		this._wsServer = this._register(
+			new WSServer(_reporter, _endpointManager, _connection)
+		);
+
+		const notUserDataDirChange = function (file: vscode.Uri): boolean {
+			return (
+				file.scheme != UriSchemes.vscode_userdata &&
+				(!_userDataDir || !PathUtil.PathBeginsWith(file.fsPath, _userDataDir))
+			);
+		};
+
+		this._register(
+			this._httpServer.onNewReqProcessed((e) => {
+				this._serverTaskProvider.sendServerInfoToTerminal(
+					e,
+					this._connection.workspace
+				);
+			})
+		);
+
+		this._register(
+			vscode.workspace.onDidChangeTextDocument((e) => {
+				if (
+					e.contentChanges &&
+					e.contentChanges.length > 0 &&
+					(e.document.uri.scheme == UriSchemes.file ||
+						e.document.uri.scheme == UriSchemes.untitled) &&
+					this._reloadOnAnyChange
+				) {
+					this._wsServer.refreshBrowsers();
+				}
+			})
+		);
+
+		this._register(
+			this._watcher.onDidChange((e) => {
+				if (this._reloadOnSave && notUserDataDirChange(e)) {
+					this._wsServer.refreshBrowsers();
+				}
+			})
+		);
+
+		this._register(
+			this._watcher.onDidDelete((e) => {
+				if (
+					(this._reloadOnAnyChange || this._reloadOnSave) &&
+					notUserDataDirChange(e)
+				) {
+					this._wsServer.refreshBrowsers();
+				}
+			})
+		);
+
+		this._register(
+			this._watcher.onDidCreate((e) => {
+				if (
+					(this._reloadOnAnyChange || this._reloadOnSave) &&
+					notUserDataDirChange(e)
+				) {
+					this._wsServer.refreshBrowsers();
+				}
+			})
+		);
+
+		this._register(
+			this._httpServer.onNewReqProcessed((e) => {
+				this._onNewReqProcessed.fire(e);
+			})
+		);
+
+		this._register(
+			this._wsServer.onConnected(() => {
+				this._wsConnected = true;
+				if (this._wsConnected && this._httpConnected) {
+					this._connected();
+				}
+			})
+		);
+
+		this._register(
+			this._httpServer.onConnected((e) => {
+				this._httpConnected = true;
+				if (this._wsConnected && this._httpConnected) {
+					this._connected();
+				}
+			})
+		);
+		this._connection.onConnected((e) => {
+			this._serverTaskProvider.serverStarted(
+				e.httpURI,
+				ServerStartedStatus.JUST_STARTED,
+				this._connection.workspace
+			);
+
+			if (this._pendingLaunchInfo) {
+				if (this._pendingLaunchInfo.external) {
+					this._previewManager.launchFileInExternalBrowser(
+						this._pendingLaunchInfo.file,
+						this._pendingLaunchInfo.relative,
+						this._pendingLaunchInfo.debug,
+						this._connection
+					);
+				} else {
+					this._previewManager.launchFileInEmbeddedPreview(
+						this._pendingLaunchInfo.file,
+						this._pendingLaunchInfo.relative,
+						this._pendingLaunchInfo.panel,
+						this._connection
+					);
+				}
+
+				this._pendingLaunchInfo = undefined;
+			}
+		});
+		vscode.commands.executeCommand('setContext', LIVE_PREVIEW_SERVER_ON, false);
```

Comment:
This is a global context key, and you are writing to it from multiple ServerGrouping objects, that could cause issues. It will be easiest to get right if a global context key is written to from a global object.